copy/copy.go

```diff
@@ -806,7 +916,7 @@ func computeDiffID(stream io.Reader, decompressor compression.DecompressorFunc)
 // and returns a complete blobInfo of the copied blob.
 func (c *copier) copyBlobFromStream(ctx context.Context, srcStream io.Reader, srcInfo types.BlobInfo,
```

Comment:
This is now far too big to be a single function. I don’t know what’s a _good_ way to restructure it yet (basically a pipeline of the `// ===`-demarcated `Reader` transformations, each one in a single function?), but this is no longer maintainable.

And we really need to have tests for at least individual pieces of this process, if not all the combinations.